basket/news/management/commands/process_fxa_queue.py

```diff
@@ -0,0 +1,90 @@
+    snitch_delay = 300  # 5 min
+    snitch_last_timestamp = 0
+    snitch_id = settings.FXA_EVENTS_SNITCH_ID
+
+    def snitch(self):
+        if not self.snitch_id:
+            return
+
+        time_since = int(time() - self.snitch_last_timestamp)
+        if time_since > self.snitch_delay:
+            requests.post('https://nosnch.in/{}'.format(self.snitch_id))
+            self.snitch_last_timestamp = time()
+
+    def handle(self, *args, **options):
+        if not settings.FXA_EVENTS_ACCESS_KEY_ID:
+            raise CommandError('AWS SQS Credentials not configured')
+
+        sqs = boto3.resource('sqs',
+                             region_name=settings.FXA_EVENTS_QUEUE_REGION,
+                             aws_access_key_id=settings.FXA_EVENTS_ACCESS_KEY_ID,
+                             aws_secret_access_key=settings.FXA_EVENTS_SECRET_ACCESS_KEY)
+        queue = sqs.Queue(settings.FXA_EVENTS_QUEUE_URL)
+
+        try:
+            # Poll for messages indefinitely.
+            while True:
+                self.snitch()
+                msgs = queue.receive_messages(WaitTimeSeconds=settings.FXA_EVENTS_QUEUE_WAIT_TIME,
+                                              MaxNumberOfMessages=10)
+                for msg in msgs:
+                    if not (msg and msg.body):
+                        continue
+
+                    statsd.incr('fxa.events.message.received')
+                    try:
+                        data = json.loads(msg.body)
+                        event = json.loads(data['Message'])
+                    except ValueError as e:
+                        # body was not JSON
+                        statsd.incr('fxa.events.message.json_error')
+                        sentry_client.captureException(data={'extra': {'msg.body': msg.body}})
+                        print('ERROR:', e, '::', msg.body)
+                        msg.delete()
+                        continue
+
+                    event_type = event.get('event', '__NONE__')
+                    statsd.incr('fxa.events.message.received.{}'.format(event_type))
+                    if event_type not in FXA_EVENT_TYPES:
+                        statsd.incr('fxa.events.message.received.{}.IGNORED'.format(event_type))
+                        # we can safely remove from the queue message types we don't need
+                        # this keeps the queue from filling up with old messages
+                        msg.delete()
+                        continue
+
+                    try:
+                        FXA_EVENT_TYPES[event_type].delay(event)
+                    except Exception:
+                        # something's wrong with the queue. try again.
+                        statsd.incr('fxa.events.message.queue_error')
+                        sentry_client.captureException(tags={'action': 'retried'})
+                        continue
+
+                    statsd.incr('fxa.events.message.success')
+                    msg.delete()
```

Comment:
Would it be useful to delete the msg just before `et_task` completes to ensure that we don't miss any msgs in case salesforge is down for example?

Reply:
I was thinking we could rely on celery retries for those cases, but it is possible to lose some if it's down for a long time. I'm not currently passing the message object to the task (not sure how well it'd go on the celery queue), but I could probably figure out how to delete the message from the task instead of the consumer; I'm just not sure it's worth going to that trouble. What do you think?

The other option is to just not use celery tasks for this and to just record the messages as they come in from the queue consumer process. I didn't do that because celery gives us retries and much more scalability, but it could work better in the case of message deletion timing.

Reply:
I hear you, I don't know if that's worth the trouble either. 

Actually re-reading this made me think: Do you plan to run this command periodically? If yes how will it get triggered?

Reply:
No, it just runs indefinitely via that infinite loop. Same as the `process_donations_quqeue` command.